xc/device.ml

```diff
@@ -607,13 +607,39 @@ let qemu_media_change ~xs device _type params =
 
 	if is_upstream_qemu device.frontend.domid
 	then begin
+		let open Qmp in
 		let cd = "ide1-cd1" in
-		let qmp_cmd =
-			if params = ""
-			then (Qmp.Command(None, Qmp.Eject (cd, Some true)))
-			else (Qmp.Command(None, Qmp.Change (cd, params, None))) in
-		qmp_write device.frontend.domid qmp_cmd
-	end
+		match params with
+		| "" ->
+			let qmp_cmd = Command(None, Eject (cd, Some true)) in
+			qmp_write device.frontend.domid qmp_cmd
+		| _ ->
+			let c = Qmp_protocol.connect (Printf.sprintf "/var/run/xen/qmp-libxl-%d" device.frontend.domid) in
+			finally
+				(fun () ->
+					try
+						let buf = " " in
+						let fd_of_c = Qmp_protocol.to_fd c in
+						let fd_of_cd = Unix.openfile params [ Unix.O_RDONLY ] 0o640 in
```

Comment:
Sorry, I'm wrong. We should close the file descriptor in toolstack, because QEMU will dup a new fd.